File: src-tauri/src/models/condition.rs
use super::FieldNameEnum;
use serde::{Deserialize, Serialize};
use ts_rs::TS;

#[derive(Serialize, Deserialize, Debug, PartialEq, TS)]
#[ts(export, export_to = "../src/models/db/db_Condition.ts")]
#[serde(rename = "db_Condition")]
pub struct Condition {
    pub name: String,
    pub description: Option<String>,
    #[serde(rename = "maleMating")]
    pub male_mating: Option<i64>,
    pub lethal: Option<bool>,
    #[serde(rename = "femaleSterile")]
    pub female_sterile: Option<bool>,
    pub arrested: Option<bool>,
    #[serde(rename = "maturationDays")]
    pub maturation_days: Option<f64>,
}

// This whole section is necessary since SQLite has no boolean type.
// Like C, it's just integers with values 0 and 1. We want to cast all of those to booleans.

impl From<ConditionDb> for Condition {
    fn from(item: ConditionDb) -> Self {
        Condition {
            name: item.name,
            description: item.description,
            male_mating: item.male_mating,
            lethal: item.lethal.map(|v| v == 1),
            female_sterile: item.female_sterile.map(|v| v == 1),
            arrested: item.arrested.map(|v| v == 1),
            maturation_days: item.maturation_days,
        }
    }
}

#[derive(Serialize, Deserialize, Debug, sqlx::FromRow)]
pub struct ConditionDb {
    pub name: String,
    pub description: Option<String>,
    pub male_mating: Option<i64>,
    // bool
    pub lethal: Option<i64>,
    // bool
    pub female_sterile: Option<i64>,
    // bool
    pub arrested: Option<i64>,
    pub maturation_days: Option<f64>,
}

#[derive(Serialize, Deserialize, Debug, Hash, PartialEq, Eq, TS)]
#[ts(export, export_to = "../src/models/db/filter/db_ConditionFieldName.ts")]
pub enum ConditionFieldName {
    Name,
    Description,
    MaleMating,
    Lethal,
    FemaleSterile,
    Arrested,
    MaturationDays,
}

impl FieldNameEnum for ConditionFieldName {
    fn get_col_name(self: &ConditionFieldName) -> String {
        match self {
            ConditionFieldName::Name => "name".to_owned(),
            ConditionFieldName::Description => "description".to_owned(),
            ConditionFieldName::MaleMating => "male_mating".to_owned(),
            ConditionFieldName::Lethal => "lethal".to_owned(),
            ConditionFieldName::FemaleSterile => "female_sterile".to_owned(),
            ConditionFieldName::Arrested => "arrested".to_owned(),
            ConditionFieldName::MaturationDays => "maturation_days".to_owned(),
        }
    }
}

File: src/models/frontend/Condition.ts
import { db_Condition } from 'models/db/db_Condition';
import { AffectedTraits } from 'models/frontend/Phenotype';

interface ICondition extends AffectedTraits {
  name: string;
  description?: string;
}

export class Condition {
  name: string = '';
  description?: string;
  maleMating?: number;
  lethal?: boolean;
  femaleSterile?: boolean;
  arrested?: boolean;
  maturationDays?: number;

  constructor(fields: ICondition) {
    Object.assign(this, fields);
  }

  static createFromRecord(record: db_Condition): Condition {
    return new Condition({
      name: record.name,
      description: record.description ?? undefined,
      maleMating:
        record.maleMating !== null ? Number(record.maleMating) : undefined,
      lethal: record.lethal ?? undefined,
      femaleSterile: record.femaleSterile ?? undefined,
      arrested: record.arrested ?? undefined,
      maturationDays: record.maturationDays ?? undefined,
    });
  }

  generateRecord = (): db_Condition => {
    return {
      name: this.name,
      description: this.description ?? null,
      maleMating:
        this.maleMating !== undefined ? BigInt(this.maleMating) : null,
      lethal: this.lethal ?? null,
      femaleSterile: this.femaleSterile ?? null,
      arrested: this.arrested ?? null,
      maturationDays: this.maturationDays ?? null,
    };
  };
}
